File: test/test_call_lua.lua
require 'torch'
require 'nn'

local TestCallLua = torch.class('TestCallLua')

function TestCallLua:__init(someName)
  print('TestCallLua:__init(', someName, ')')
  assert(someName == 'green')
  self.someName = someName
end

function TestCallLua:getName()
  return self.someName
end

function TestCallLua:getOut(inTensor, outSize, kernelSize)
  local inSize = inTensor:size(3)
  local m = nn.TemporalConvolution(inSize, outSize, kernelSize)
  m:float()
  local out = m:forward(inTensor)
  print('out from lua', out)
  return out
end

function TestCallLua:addThree(inTensor)
  local outTensor = inTensor + 3
  return outTensor
end

function TestCallLua:printTable(sometable, somestring, table2)
  for k, v in pairs(sometable) do
    print('TestCallLua:printTable ', k, v)
  end
  assert(sometable['color'] == 'red')
  assert(sometable['weather'] == 'sunny')
  assert(sometable['anumber'] == 10)
  assert(sometable['afloat'] - 1.234 < 1e-7)
  assert(somestring == 'mistletoe')
  print('somestring', somestring)
  for k, v in pairs(table2) do
    print('TestCallLua table2 ', k, v)
    if k == 'row1' then
        assert(v == 'col1')
    elseif k == 'meta' then
        assert(v == 'data')
    end
  end
  return {bear='happy', result=12.345, foo='bar'}
end

function TestCallLua:getList(somelist)
    return {3.1415, 'Lua', 123}
end

function TestCallLua:modifyList(somelist)
  assert(type(somelist) == 'table')
  print('list in lua:')
  for i = 1, #somelist do
    print(i, somelist[i])
  end
  assert((somelist[1] - 3.1415) < 1e-7)
  assert(somelist[2] == '~Python\\omega')
  assert(somelist[3] == 42)

  table.insert(somelist, 'Lorem Ipsum')
  return somelist
end

File: test/test_call_lua.py
import PyTorch
import PyTorchHelpers
import numpy as np


def test_call_lua():
    TestCallLua = PyTorchHelpers.load_lua_class('test/test_call_lua.lua', 'TestCallLua')

    batchSize = 2
    numFrames = 4
    inSize = 3
    outSize = 3
    kernelSize = 3

    luabit = TestCallLua('green')
    print(luabit.getName())
    assert luabit.getName() == 'green'
    print('type(luabit)', type(luabit))
    assert str(type(luabit)) == '<class \'PyTorchLua.TestCallLua\'>'

    np.random.seed(123)
    inTensor = np.random.randn(batchSize, numFrames, inSize).astype('float32')
    luain = PyTorch.asFloatTensor(inTensor)

    luaout = luabit.getOut(luain, outSize, kernelSize)

    outTensor = luaout.asNumpyTensor()
    print('outTensor', outTensor)
    # I guess we just assume if we got to this point, with no exceptions, then thats a good thing...
    # lets add some new test...

    outTensor = luabit.addThree(luain).asNumpyTensor()
    assert isinstance(outTensor, np.ndarray)
    assert inTensor.shape == outTensor.shape
    assert np.abs((inTensor + 3) - outTensor).max() < 1e-4

    res = luabit.printTable({'color': 'red', 'weather': 'sunny', 'anumber': 10, 'afloat': 1.234}, 'mistletoe', {
        'row1': 'col1', 'meta': 'data'})
    print('res', res)
    assert res == {'foo': 'bar', 'result': 12.345, 'bear': 'happy'}

    reslist = luabit.getList(['blue', 42, r'\omega'])
    restuple = luabit.getList((3.1415, r'~Python~', 'Torch', 1.2e-4))
    assert reslist == restuple == {1: 'lorem', 2: 'ipsum', 3: 42, 4: 2.1}